perf/memtx.cc

```diff
@@ -0,0 +1,188 @@
+					   memtx_alloc_factor);
+		memtx->state = MEMTX_OK;
+
+		::port_init();
+
+		struct space_opts space_opts;
+		::space_opts_create(&space_opts);
+		space_opts.group_id = GROUP_LOCAL;
+		space_opts.is_temporary = true;
+		space_def = ::space_def_new(sid, uid, 0, "perf",
+					    std::strlen("perf"), "memtx",
+					    std::strlen("memtx"), &space_opts,
+					    &field_def_default, 0);
+		RLIST_HEAD(key_list);
+		struct index_opts idx_opts;
+		::index_opts_create(&idx_opts);
+		std::uint32_t fields[] = { 0 };
+		std::uint32_t types[] = { FIELD_TYPE_UNSIGNED };
+		key_def = ::box_key_def_new(fields, types, 1);
+		tree_idx_def = ::index_def_new(sid, tree_iid, "pk",
+					       std::strlen("pk"), TREE,
+					       &idx_opts, key_def, nullptr);
+		::index_def_list_add(&key_list, tree_idx_def);
+		space = ::memtx_space_new(memtx, space_def, &key_list);
+		::space_cache_replace(nullptr, space);
+		::space_add_primary_key(space);
+
+		for (auto &k: key_set)
+			::box_insert(sid, k.first, k.second, nullptr);
+	}
+
+	void
+	TearDown(const benchmark::State &) override
+	{
+		::space_def_delete(space_def);
+		::key_def_delete(key_def);
+
+		::session_free();
+		::user_cache_free();
+		::space_cache_destroy();
+		::tuple_free();
+
+		memtx->base.vtab->shutdown(&memtx->base);
+
+		::fiber_free();
+		::memory_free();
+	}
+
+private:
+	void
+	generate_key_set()
+	{
+		for (std::size_t i = 0; i < key_set.size(); ++i) {
+			key_set[i].second =
+				::mp_encode_array(key_set[i].first, 1);
+			key_set[i].second =
+				::mp_encode_uint(key_set[i].second, i);
+		}
+	}
+
+	static constexpr unsigned int rand_seed = 0;
+
+	static constexpr std::uint32_t memtx_objsize_min = 16;
+	static constexpr unsigned int memtx_granularity = 8;
+	static constexpr float memtx_alloc_factor = 1.1;
+	struct memtx_engine *memtx;
+
+	struct space_def *space_def;
+	struct key_def *key_def;
+	struct index_def *tree_idx_def;
+	struct space *space;
+
+	static constexpr std::size_t key_size_max = 5 + 9;
+	static constexpr std::size_t key_set_size = 1 << 10;
+	std::array<std::pair<char[key_size_max], char *>, key_set_size> key_set;
+	static constexpr std::size_t key_subset_size = key_set_size / 2;
+
+protected:
+	void
+	generate_key_subset()
+	{
+		for(auto &k: key_subset)
+			k = key_set[std::rand() % key_set.size()];
+	}
+
+	static constexpr std::uint32_t uid = 0;
+	static constexpr std::uint32_t sid = 0;
+	static constexpr std::uint32_t tree_iid = 0;
+
+	std::array<std::pair<char *, char *>, key_subset_size> key_subset;
+
+	static constexpr std::uint32_t select_lim = 4294967295;
+	static constexpr std::size_t empty_key_size_max = 5;
+	std::pair<char[empty_key_size_max], char *> empty_key;
+};
+
+BENCHMARK_F(MemtxFixture, TreeGetRandomExisting)
+(benchmark::State &state)
+{
+	std::int64_t iterations_count = 0;
+	for (auto _: state) {
+		for (auto &k: key_subset) {
+			++iterations_count;
+			struct tuple *t;
+			::box_index_get(sid, tree_iid, k.first, k.second, &t);
```

Comment:
I think `DoNotOptimize` must be added here. Or better check the return value and panic if it's non-zero.

Reply:
Fixed: wrapped it with `DoNotOptimize`.

I believe checking the return value would affect the measurement.